File: learn_android/code/JAVA/SMSdemo/app/src/main/java/com/hkc/smsdemo/MainActivity.java
package com.hkc.smsdemo;

import android.Manifest;
import android.content.pm.PackageManager;
import android.os.Bundle;
import android.telephony.SmsManager;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.app.ActivityCompat;
import androidx.core.content.ContextCompat;

public class MainActivity extends AppCompatActivity {
    final int SEND_SMS_CODE = 100;
    Button sendButton;
    EditText et_message,et_phone;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        sendButton = findViewById(R.id.sendButton);
        et_message = findViewById(R.id.et_message);
        et_phone = findViewById(R.id.et_phone);




        sendButton.setOnClickListener(new View.OnClickListener() {
        @Override
        public void onClick(View view) {
            int permission = ContextCompat.checkSelfPermission(MainActivity.this, Manifest.permission.SEND_SMS);
            if ( permission == PackageManager.PERMISSION_GRANTED ){
                sendMseeage();
            }else{
                requestPermissions();

            }

        }


    });

    }

    private void requestPermissions() {
        ActivityCompat.requestPermissions(MainActivity.this,new String[]{Manifest.permission.SEND_SMS},SEND_SMS_CODE);

    }

    private void sendMseeage() {
        //
        String sPhone = et_phone.getText().toString().trim();
        String sMessage = et_message.getText().toString().trim();
        //
        if(!sPhone.equals("")  && !sMessage.equals("")){
            //
            //
            SmsManager smsManager = SmsManager.getDefault();
            //Send Text Message
            smsManager.sendTextMessage(sPhone,null,sMessage,null,null);
            //
            Toast.makeText(getApplicationContext(),"SMS send Successful!",Toast.LENGTH_LONG).show();

        }else{
            //
            Toast.makeText(getApplicationContext(),"Enter value first",Toast.LENGTH_LONG).show();
        }
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);
        //
        if (requestCode == SEND_SMS_CODE && grantResults.length> 0 && grantResults[0] == PackageManager.PERMISSION_GRANTED){
            sendMseeage();
        }else{
//            requestPermissions();
            Toast.makeText(getApplicationContext(), "Permission Denied!", Toast.LENGTH_LONG).show();
        
        }
    }
}

File: learn_android/code/demo2/VideoRecoder/app/src/main/java/com/hkc/videorecorder/MainActivity.kt
package com.hkc.videorecorder

import android.content.ActivityNotFoundException
import android.content.Intent
import android.graphics.Bitmap
import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.provider.MediaStore
import android.widget.Button
import android.widget.*

class MainActivity : AppCompatActivity() {

    private var angle = 0f
    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        //識別返回對象及結果
        if (requestCode == 0 && resultCode == RESULT_OK){
            //取得影像並顯示於ImageView
            val image = data?.extras?.get("data") ?: return
            findViewById<ImageView>(R.id.imageView).setImageBitmap(image as Bitmap)

        }

    }



    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        findViewById<Button>(R.id.btn_photo).setOnClickListener {
        //建立一個要進行影像獲取的Intent物件
            val intent = Intent(MediaStore.ACTION_IMAGE_CAPTURE)
            //用try-catch避免產生例外，若產生則顯示Toast
            try{
                startActivityForResult(intent,0)        //發送Intent


            }catch (e: ActivityNotFoundException){
                Toast.makeText(this,"此裝置無相機應用程式",Toast.LENGTH_LONG).show()

            }


        }
        findViewById<Button>(R.id.btn_rotate).setOnClickListener {
            angle += 90f        //原本角度再加上90度
            findViewById<ImageView>(R.id.imageView).rotation = angle    //使ImageView旋轉
        }


    }
}